### PR TITLE
Implemented Photon Geyser and Shell Side Arm

### PR DESCRIPTION
**Pull request by LOuroboros**
## Description
Supersedes #1523 because its author can't mantain it any longer, and closes both, #515 and #1090.

## **Discord contact info**
Lunos#4026

### CODE REVIEW

File: src/battle_script_commands.c [hunk 1]
```diff
@@ -8980,6 +8981,57 @@ static void Cmd_various(void)
...
+
+        if (attStat > spaStat)
+            gSwapDamageCategory = TRUE;
+        break;
+    }
+    case VARIOUS_SHELL_SIDE_ARM_CHECK: // 0% chance GameFreak actually checks this way according to DaWobblefet, but this is the only functional explanation at the moment
+    {
+        u32 attStat = gBattleMons[gBattlerAttacker].attack;
+        u8 atkStage = gBattleMons[gBattlerAttacker].statStages[STAT_ATK];
+        u32 attStatDef = gBattleMons[gBattlerTarget].attack;
```
Comment (BuffelSaft) [inline src/battle_script_commands.c:9005]: This is checking the wrong stats - should be the target's defenses, not its attack stats

**Comment by BuffelSaft** [comment]
Shell Side Arm is checking the wrong stats to determine its damage type. It should checking be the target's defensive stats, not its offensive stats. From Smogon: "This move becomes a physical attack that makes contact if the value of ((((2 * the user's level / 5 + 2) * 90 * X) / Y) / 50), where X is the user's Attack stat and Y is the target's Defense stat, is greater than the same value where X is the user's Special Attack stat and Y is the target's Special Defense stat."

But `case VARIOUS_SHELL_SIDE_ARM_CHECK` is doing this:
```        
u32 attStat = gBattleMons[gBattlerAttacker].attack;
u8 atkStage = gBattleMons[gBattlerAttacker].statStages[STAT_ATK];
u32 attStatDef = gBattleMons[gBattlerTarget].attack;
```

`u32 attStatDef` should be the target's defense stat. I'd suggest renaming this variable to something more descriptive like `targetDefStat`, and changing `atkStage` to `statStage` as it's not just used to store the attack stat stage.

File: src/battle_script_commands.c [hunk 2]
```diff
@@ -9005,6 +9005,57 @@ static void Cmd_various(void)
...
+        u32 targetSpDefStat = gBattleMons[gBattlerTarget].spDefense;
+        u8 statStage;
+        u32 physical;
+        u32 special;
+
+        statStage = gBattleMons[gBattlerAttacker].statStages[STAT_ATK];
+        attackerAtkStat *= gStatStageRatios[statStage][0];
+        attackerAtkStat /= gStatStageRatios[statStage][1];
+
+        statStage = gBattleMons[gBattlerTarget].statStages[STAT_ATK];
```
Comment (BuffelSaft) [inline src/battle_script_commands.c:9039]:
```suggestion
        statStage = gBattleMons[gBattlerTarget].statStages[STAT_DEF];
```
I think this should be getting the target's defense stat stages

File: src/battle_script_commands.c [hunk 3]
```diff
@@ -9005,6 +9005,57 @@ static void Cmd_various(void)
...
+        targetDefStat *= gStatStageRatios[statStage][0];
+        targetDefStat /= gStatStageRatios[statStage][1];
+
+        physical = ((((2 * gBattleMons[gBattlerAttacker].level / 5 + 2) * gBattleMoves[gCurrentMove].power * attackerAtkStat) / targetDefStat) / 50);
+
+        statStage = gBattleMons[gBattlerAttacker].statStages[STAT_SPATK];
+        attackerSpAtkStat *= gStatStageRatios[statStage][0];
+        attackerSpAtkStat /= gStatStageRatios[statStage][1];
+
+        statStage = gBattleMons[gBattlerTarget].statStages[STAT_SPATK];
```
Comment (BuffelSaft) [inline src/battle_script_commands.c:9049]:
```suggestion
        statStage = gBattleMons[gBattlerTarget].statStages[STAT_SPDEF];
```
Same thing here, should be the target's special defense stages

**Review by BuffelSaft — COMMENTED** [review]
The new var names are much better, makes the code easier to understand. I'm having trouble getting Shell Side arm to switch its category properly but can't see why yet, will keep looking.

**Comment by BuffelSaft** [comment]
I think these are both working now, tested against various stat stage combinations, light screen, reflect, fur coat, contact abilities, etc.

**Review by BuffelSaft — APPROVED** [review]
Oops, ignore this

**Comment by Skeli789** [comment]
Sorry, I might be a bit late to this, but I happened to be going through some of these pull requests and noticed there are a few problems here.

1. The way Shell Side Arm was implemented is inconsistent with how it actually works in the official games. It's supposed to choose the damage category at the beginning of the turn, and not mid-turn. See [here](https://www.smogon.com/forums/threads/sword-shield-battle-mechanics-research.3655528/post-8508168).
2. It doesn't take into account Reflect or Light Screen. See link in 1.
3. It doesn't take into account Wonder Room. See [here](https://www.smogon.com/forums/threads/sword-shield-battle-mechanics-research.3655528/post-8707046). Similar things like Body Press and Download may be implemented incorrectly in other places in your battle engine, but I haven't checked.
4. The AI would have no idea how to use it properly due to this implementation. If you hooked it directly into the damage calculator, it would work much better.

**Comment by BuffelSaft** [comment]
No worries, thanks for bringing this up! I wasn't aware Shell Side Arm worked that way, it's hard to find detailed information on it. Wonder Room will likely need changes as I don't think there's any special handling for those cases right now. I assume moving `gSwapDamageCategory` into `gBattleStruct` will fix the link battle issue?

AFAIK there's no particular reason for the global variable use, that needs to be looked into!

**Comment by Skeli789** [comment]
> it's hard to find detailed information on it

When in doubt, it's best to look through the Smogon [thread](https://www.smogon.com/forums/threads/sword-shield-battle-mechanics-research.3655528/) I linked before. It's a lot more detailed than Bulbapedia, and you can easily search for mechanics there.

> Wonder Room will likely need changes as I don't think there's any special handling for those cases right now.

You can clearly tell it's incredibly buggy in the official games, but unfortunately, I doubt GF will ever fix those bugs, making them official "features."

>  I assume moving gSwapDamageCategory into gBattleStruct will fix the link battle issue?

No, you'd need to send the data over in ``BtlController_EmitMoveAnimation`` and retrieve it before/during the animation. 

> AFAIK there's no particular reason for the global variable use

In that case, unless you need to access a certain variable in a battle script,  I highly recommend that all battle-related memory values be placed in ``gBattleStruct`` (for the reason I mentioned before).

**Comment by BuffelSaft** [comment]
> When in doubt, it's best to look through the Smogon [thread](https://www.smogon.com/forums/threads/sword-shield-battle-mechanics-research.3655528/) I linked before. It's a lot more detailed than Bulbapedia, and you can easily search for mechanics there.

Will definitely do that in future!

> You can clearly tell it's incredibly buggy in the official games, but unfortunately, I doubt GF will ever fix those bugs, making them official "features."

Mm, there seem to be a lot of those (Sheer Force in general is a disaster). It's almost worth having bug fix configs for some of them.

> In that case, unless you need to access a certain variable in a battle script, I highly recommend that all battle-related memory values be placed in `gBattleStruct` (for the reason I mentioned before).

Agreed!

**Pull request merged**